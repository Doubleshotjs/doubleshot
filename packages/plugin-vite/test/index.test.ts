import path from 'node:path'
import { execa } from '@esm2cjs/execa'
import fs from 'fs-extra'
import { afterAll, beforeAll, describe, expect, it } from 'vitest'

const mockDir = path.resolve(__dirname, './mock')

async function installDeps(cwd: string) {
  const { stdout, stderr } = await execa(
    'pnpm',
    ['install', '--no-lockfile', '--ignore-workspace'],
    {
      cwd,
    },
  )

  const logs = stdout + stderr
  return logs
}

function remove() {
  fs.removeSync(path.resolve(mockDir, 'dist'))
}

async function run(mode: 'dev' | 'build') {
  const { stdout, stderr } = await execa(
    'npm',
    ['run', mode],
    {
      cwd: mockDir,
    },
  )

  const logs = stdout + stderr
  return logs
}

beforeAll(async () => {
  remove()
  await installDeps(mockDir)
}, 10 * 60 * 1000)

afterAll(() => {
  remove()
})

describe('doubleshot Vite Plugin', () => {
  it('should run dev mode', async () => {
    const logs = await run('dev')

    expect(logs).toContain('Development')
  })

  describe('plugin\'s config should be overridden by `configureForMode`', async () => {
    it('should match snapshot in development mode', async () => {
      const devLogs = await run('dev')
      const devResult = fs.readFileSync(path.resolve(mockDir, 'dist/main/index.js'), 'utf8')
      expect(devLogs).toContain('override config for development')
      expect(devResult).toMatchSnapshot()
    })

    it('should match snapshot in production mode', async () => {
      const prodLogs = await run('build')
      const prodResult = fs.readFileSync(path.resolve(mockDir, 'dist/main/index.js'), 'utf8')
      expect(prodLogs).toContain('override config for production')
      expect(prodResult).toMatchSnapshot()
    })
  })

  it('should also inject preload', async () => {
    const logs = await run('dev')

    expect(logs).toContain('hello, this is a msg through preload file')
  })

  it('should build electron app if "electron.build" is set', async () => {
    const logs = await run('build')

    expect(logs).toContain('Start electron build')
    expect(logs).toContain('Build succeeded')
    expect(fs.existsSync(path.resolve(mockDir, 'dist/electron'))).toBe(true)
  }, 10 * 60 * 1000)
})
