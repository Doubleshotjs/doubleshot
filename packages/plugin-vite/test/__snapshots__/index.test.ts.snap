// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`doubleshot Vite Plugin > plugin's config should be overridden by \`configureForMode\` > should match snapshot in development mode 1`] = `
""use strict";

// src/index.ts
var import_node_path = require("path");
var import_electron = require("electron");
function createWindow() {
  const isDev = !import_electron.app.isPackaged;
  const win = new import_electron.BrowserWindow({
    width: 800,
    height: 600,
    webPreferences: {
      contextIsolation: true,
      preload: (0, import_node_path.join)(__dirname, "./preload.js")
    }
  });
  import_electron.ipcMain.handle("msg", (_, msg) => {
    console.log(msg);
    import_electron.app.quit();
  });
  const URL = isDev ? "http://127.0.0.1:12345/" : \`file://\${(0, import_node_path.join)(import_electron.app.getAppPath(), "dist/renderer/index.html")}\`;
  win.loadURL(URL);
}
import_electron.app.whenReady().then(() => {
  createWindow();
});
import_electron.app.on("window-all-closed", () => {
  if (process.platform !== "darwin")
    import_electron.app.quit();
});
"
`;

exports[`doubleshot Vite Plugin > plugin's config should be overridden by \`configureForMode\` > should match snapshot in production mode 1`] = `
""use strict";var import_node_path=require("path");var import_electron=require("electron");function createWindow(){const isDev=!import_electron.app.isPackaged;const win=new import_electron.BrowserWindow({width:800,height:600,webPreferences:{contextIsolation:true,preload:(0,import_node_path.join)(__dirname,"./preload.js")}});import_electron.ipcMain.handle("msg",(_,msg)=>{console.log(msg);import_electron.app.quit()});const URL=isDev?process.env.DS_RENDERER_URL:\`file://\${(0,import_node_path.join)(import_electron.app.getAppPath(),"dist/renderer/index.html")}\`;win.loadURL(URL)}import_electron.app.whenReady().then(()=>{createWindow()});import_electron.app.on("window-all-closed",()=>{if(process.platform!=="darwin")import_electron.app.quit()});
"
`;

exports[`doubleshot Vite Plugin > should build electron app if "electron.build" is set 1`] = `
""use strict";var import_node_path=require("path");var import_electron=require("electron");function createWindow(){const isDev=!import_electron.app.isPackaged;const win=new import_electron.BrowserWindow({width:800,height:600,webPreferences:{contextIsolation:true,preload:(0,import_node_path.join)(__dirname,"./preload.js")}});import_electron.ipcMain.handle("msg",(_,msg)=>{console.log(msg);import_electron.app.quit()});const URL=isDev?process.env.DS_RENDERER_URL:\`file://\${(0,import_node_path.join)(import_electron.app.getAppPath(),"dist/renderer/index.html")}\`;win.loadURL(URL)}import_electron.app.whenReady().then(()=>{createWindow()});import_electron.app.on("window-all-closed",()=>{if(process.platform!=="darwin")import_electron.app.quit()});
"
`;
